src/Compilers/VisualBasic/Portable/Symbols/ConstraintsHelper.vb

```diff
@@ -610,6 +649,140 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
+                                                    tupleNamesCheckKind As TupleNamesCheckKind) As Boolean
+
+            ' The skip list is used to avoid duplicate checks. It also avoid circular checks, as in `Class Derived Inheirts Base(Of Derived)`.
+            Dim skipList = PooledHashSet(Of TypeSymbol).GetInstance()
+            Dim result = CheckTupleNamesConstraints(type, diagnosticsBuilder, tupleNamesCheckKind, skipList)
+            skipList.Free()
+            Return result
+        End Function
+
+        Private Function CheckTupleNamesConstraints(type As TypeSymbol,
+                                                    <[In], Out> ByRef diagnosticsBuilder As ArrayBuilder(Of TypeParameterDiagnosticInfo),
+                                                    tupleNamesCheckKind As TupleNamesCheckKind,
+                                                    skipList As HashSet(Of TypeSymbol)) As Boolean
+            ' Note this only checks for tuple name errors introduced by referencing this type.
+            '
+            ' So if the type does not include any tuples, its reference cannot introduce an error.
+            ' But its declaration could. For instance:
+            '     `Class Error Inherits Base Implements I(Of (notA As Integer, notB As Integer))`
+            '
+            ' Similarly, if the base type doesn't use type arguments from the type, then the reference of this type cannot introduce an error.
+            ' But the base type may already have one (reported elsewhere). For instance:
+            '     `Class C Inherits BaseCanBeSkipped(Of Integer) Implements ITest`
+            '     `Class D Inherits BaseNeedsChecking(Of T) Implements ITest`
+            Debug.Assert(type IsNot Nothing And skipList IsNot Nothing)
+            If skipList.Contains(type) OrElse Not type.ContainsTuple() Then
+                Return True
+            End If
+
+            Select Case tupleNamesCheckKind
+                Case TupleNamesCheckKind.Shallow
+                    skipList.Add(type)
+                    If Not CheckTupleNamesConstraintsCore(type, diagnosticsBuilder) Then
+                        Return False
+                    End If
+
+                    ' Walk up the base types and apply the same check on each inheritence level.
+                    ' We can cut short if the original definition's base doesn't contain type parameters.
+                    Dim currentType = type
+                    Dim baseType = currentType.BaseTypeNoUseSiteDiagnostics
+                    Dim result = True
+
+                    While baseType IsNot Nothing AndAlso IsOrRefersToTypeParameter(currentType.OriginalDefinition.BaseTypeNoUseSiteDiagnostics)
+                        result = result And CheckTupleNamesConstraints(baseType, diagnosticsBuilder, TupleNamesCheckKind.Deep, skipList)
+                        currentType = baseType
+                        baseType = currentType.BaseTypeNoUseSiteDiagnostics
```

Comment:
Thanks, this is nasty. No, we don't handle this well.
If you add interfaces and tuples into this example (to avoid the shortcuts in the exploration), then this blows up. 
It will check `B<A<(int, int)>>`, then `B<A<A<(int, int)>>>` and so on...
I need to figure out how to protect against such explosion.